### ISSUE TITLE
Integration needs a reload when entity becomes unavailable

### ISSUE THREAD

**Issue by TroLoos** [pnbruckner/ha-composite-tracker#58]
Hi, first of all - I love the idea of this integration, I really do. Actually, I've been hoping it could become another layer of resilience for my home setup. Here's what is my current issue.

I want to use Composite Tracker to diversify home/not_home status based on more than one entity so that if one of them becomes unavailable, the status would be still reported correctly.

My current sample configuration:
```
composite:
  default_options:
    require_movement: false
  trackers:
    - name: Karol
      id: karol
      entity_id:
        - entity: device_tracker.karol_s_iphone
        - entity: binary_sensor.karol_s_iphone_wifi_at_home
```
device_tracker.karol_s_iphone - comes from Unifi integration
binary_sensor.karol_s_iphone_wifi_at_home - comes from template sensor and is update when HA companion app connects/disconnects to/from my home WiFi.

The problem is that when I disable Unifi integration / device_tracker.karol_s_iphone becomes unavailable (there are some issues ongoing with current firmware when such things happen randomly and quite often), Composite Tracker stops reacting to changes of tracker that is still available / binary_sensor.karol_s_iphone_wifi_at_home.

Are my observations correct? Would it be complicated to add logic that skips the entities that are unavailable?

**Comment by pnbruckner** [comment]
Composite already ignores `unavailable` and `unknown` states. When this happens, it is effectively remembering the last "good" state (i.e., before it became unavailable.) So, for the router-based tracker, it's remembering it is home.

By default, composite has to determine if it should use new states for non-GPS inputs (e.g., router-based trackers & binary sensors.) It always uses their state when they change to "home." However, when they change to not "home", it gets a bit complicated. Suffice it to say, the default behavior, in your scenario. means the binary sensor changing to off (while the router-based tracker is unavailable) will _not_ cause the composite to change to not "home."

To solve your problem, I would suggest enabling `all_states` for the binary sensor. I.e.,:
```yaml
composite:
  default_options:
    require_movement: false
  trackers:
    - name: Karol
      id: karol
      entity_id:
        - device_tracker.karol_s_iphone
        - entity: binary_sensor.karol_s_iphone_wifi_at_home
          all_states: true
```
Now the composite will change to not "home" when the binary sensor goes to "off". The downside is that this will always happen, even when the router-based tracker is "home" (i.e., not unavailable and indicating home.) But, maybe given your situation, this could work.

**Comment by TroLoos** [comment]
@pnbruckner thank you for a very comprehensive reply, now I learned what `all_states` really does - I thought it is a method to enable "home" state only when all trackers are at "home".

> Composite already ignores `unavailable` and `unknown` states. When this happens, it is effectively remembering the last "good" state (i.e., before it became unavailable.) So, for the router-based tracker, it's remembering it is home.

In my case (and any other resilient-related) it would be great to have another switch (like `ignore_unhealthy`) to literally ignore all non-healthy entities and use only good ones to calculate current state at most probability.

> By default, composite has to determine if it should use new states for non-GPS inputs (e.g., router-based trackers & binary sensors.) It always uses their state when they change to "home." However, when they change to not "home", it gets a bit complicated. Suffice it to say, the default behavior, in your scenario. means the binary sensor changing to off (while the router-based tracker is unavailable) will _not_ cause the composite to change to not "home."

Yes, that is exactly what I can observe.

> To solve your problem, I would suggest enabling `all_states` for the binary sensor. I.e.,:
> 
> ```yaml
> composite:
>   default_options:
>     require_movement: false
>   trackers:
>     - name: Karol
>       id: karol
>       entity_id:
>         - device_tracker.karol_s_iphone
>         - entity: binary_sensor.karol_s_iphone_wifi_at_home
>           all_states: true
> ```
> 
> Now the composite will change to not "home" when the binary sensor goes to "off". The downside is that this will always happen, even when the router-based tracker is "home" (i.e., not unavailable and indicating home.) But, maybe given your situation, this could work.

Right, but that would mean that this tracker is superior over all other trackers. This unfortunately doesn't solve the problem of resiliency - turns out that iOS companion app also sometimes likes to stop reporting, it actually happened to me today, that's why I had to investigate a little my setup and I observed that it doesn't work as I thought it would be. In this case, if it fails to update, it will ignore all other trackers (even healthy ones) and might cause wrong state.

Anyway, I will have that in mind and maybe test some other scenarios (like including GPS based tracker, I have to re-think it). But if you accept the challange and decide to look at my scenario, I would be more than grateful ;-).

**Comment by TroLoos** [comment]
Actually, I figured a method (not yet implemented) to achieve the functionality I really need.

I will create a kind of intermediate-sensor with template, that takes device_tracker as input (the one that might become unavailable) and template it to reflect the state:
- home (when entity is home)
- not_home (when entity is not_home / unavailable)

This way - my composite tracker will take only the healthy sensors into account, when calculation the state of combo-tracker. And on the other hand (because that could also be simply achieved by templating multiple entities), I will have on result site entity that is device_tracker class.

I think it should be easy to implement :-).

**Comment by pnbruckner** [comment]
I can see what you're saying. But I'm not sure the composite integration, given its design, can do anything about it. I'll have to give it some more thought.

But, in the meantime, composite isn't designed to look at the current state of all its inputs and decide "which is best." (If that's what you want, maybe the person integration will work for you, because that's basically what it does.) It is designed based on the concept that "latest is best."

So, it responds to _changes_ in its inputs, and for each change, decides whether or not to use that new information. It bases that on several things, but first and most importantly is the timestamp of the data. If the data is _older_ than the previously used data (possibly from a different input), then it is ignored. It also ignores any new state that is unavailable or unknown (which is just another way of saying, "I don't have any new data.")

Once it decides the data is "good", and newer, then it has to look at other details, e.g., what _type_ of input it is. It does this because non-GPS based inputs (router-based trackers, binary sensors, etc.) are pretty good at detecting when a device "comes home" but are notoriously bad at detecting when they "leave home". Typically, that change is based on the _lack of data_ (i.e., the device hasn't been seen for some arbitrary time period) rather than positively knowing the device has left. This is why, by default, for these types of inputs, only the "to home" transitions are used, and the "from home" transitions are ignored.

Again, the composite is based on using _new data_. The _lack of data_ (which happens in all sampling systems, which HA definitely is) is not new data. When an input becomes unavailable, that is just another case of time between good updates.

Anyway, that's the design philosophy. It does treat non-GPS based inputs specially due to their nature, but otherwise, newer is better than older.

**Comment by TroLoos** [comment]
@pnbruckner thank you once again for a very comprehensive answer, I'm sure it will be a good source of knowledge for other users that want to know a little bit more about the logic behind this integration.

I decided to stick to it instead of `person` integration - In contrast to `person`, as far as I understand, this integration takes into account all non-gps based entities so that if any of them is `home` then the result is `home`, no matter what was the last change of a group.

I achieved exactly what I needed like I described in my previous message - by converting device_tracker from router to binary_sensor, and used them as input to composite tracker. They never have `unavailable` state so they only change to `not_home`. And everything works great!

I think we can close this issue as resolved.
